program/tests/denylist_royalty_enforcement.rs

```diff
@@ -0,0 +1,1138 @@
+            assert_eq!(error, RuleSetError::NotImplemented as u32);
+        }
+        _ => panic!("Unexpected error: {:?}", err),
+    }
+}
+
+#[tokio::test]
+async fn wallet_to_prog_owned() {
+    let mut context = program_test().start_with_context().await;
+    let rule_set_addr = create_royalty_rule_set(&mut context).await;
+
+    // Create a Keypair to simulate a token mint address.
+    let mint = Keypair::new();
+
+    // Source key is a wallet.
+    let source = Keypair::new();
+
+    // Our destination key is going to be an account owned by the mpl-token-auth-rules program.
+    // Any one will do so for convenience we just use the RuleSet.
+
+    // Get on-chain account.
+    let on_chain_account = context
+        .banks_client
+        .get_account(rule_set_addr)
+        .await
+        .unwrap()
+        .unwrap();
+
+    // Account must have nonzero data to count as program-owned.
+    assert!(on_chain_account.data.iter().any(|&x| x != 0));
+
+    // Verify account ownership.
+    assert_eq!(mpl_token_auth_rules::ID, on_chain_account.owner);
+
+    let payload = Payload::from([
+        (PayloadKey::Amount.to_string(), PayloadType::Number(1)),
+        (
+            PayloadKey::Source.to_string(),
+            PayloadType::Pubkey(source.pubkey()),
+        ),
+        (
+            PayloadKey::Destination.to_string(),
+            PayloadType::Pubkey(rule_set_addr),
+        ),
+        (
+            PayloadKey::Authority.to_string(),
+            PayloadType::Pubkey(context.payer.pubkey()),
+        ),
+    ]);
+
+    let transfer_owner_operation = Operation::Transfer {
+        scenario: TransferScenario::Holder,
+    };
+
+    // Create a `validate` instruction.
+    let validate_ix = ValidateBuilder::new()
+        .rule_set_pda(rule_set_addr)
+        .mint(mint.pubkey())
+        .additional_rule_accounts(vec![
+            AccountMeta::new_readonly(source.pubkey(), false),
+            AccountMeta::new_readonly(rule_set_addr, false),
+            AccountMeta::new_readonly(context.payer.pubkey(), true),
+        ])
+        .build(ValidateArgs::V1 {
+            operation: transfer_owner_operation.to_string(),
+            payload,
+            update_rule_state: false,
+            rule_set_revision: None,
+        })
+        .unwrap()
+        .instruction();
+
+    // Validate operation.
+    let err =
+        process_failing_validate_ix!(&mut context, validate_ix, vec![], Some(ADDITIONAL_COMPUTE))
+            .await;
+
+    // Check that error is what we expect.  Program owner was not on the allow list.
+    match err {
+        solana_program_test::BanksClientError::TransactionError(
+            TransactionError::InstructionError(_, InstructionError::Custom(error)),
+        ) => {
+            assert_eq!(error, RuleSetError::ProgramOwnedListCheckFailed as u32);
+        }
+        _ => panic!("Unexpected error: {:?}", err),
+    }
+}
+
+#[tokio::test]
+async fn wallet_to_prog_owned_missing_namespace() {
```

Comment:
This test isn't really that the namespace is missing, right?  Its more like the fact that the operation is missing is why it fails.  Specifically, `Transfer:Owner` would pass if it either used the `Transfer` namespace rule, OR if it had its own explicit rule.

I think your next test called `wallet_to_prog_owned_no_default` is what I would have called missing namespace.

The test names were a bit confusing at first to me.  Maybe the names of these two tests could be updated to be more describing of the scenarios?  Or at least some comments above the `create_incomplete_royalty_rule_set` to help explain what exactly will happen when this operation gets removed.